### PR TITLE
Ansible provisioner

### PR DESCRIPTION
**Pull request by mattyjones**
REF: #2 

Replaces #1 

This will bring in ansible to provision and configure a stable build environment for packaging and/or testing the source in any supported environment.

### CODE REVIEW

**Comment by phoerious** [comment]
Does "Replaces #1" mean that #1 is obsolete? If yes, it'd probably be best to close it.

**Comment by mattyjones** [comment]
@phoerious It will replace it but I am not sure when this will be ready. I am hoping end of week to have a stable build and test env. It also depends on if people want or see the need for tools like Ansible.

I always defer to them due to personal exp and believing that everything should be treated as code and infrastructure should be immutable. While I can achieve this with shell scripts, the complexity in my mind is not worth it and we should just use tools that are designed for this process.

If there is no strong opinions about using Ansible then yes, we can close #1 and just wait for this to drop but I wanted to ensure that others who have come before also have a voice and can contribute. I don't mean to come on the scene and dictate tools and workflows if others already have stuff in place that works well.

**Comment by samrocketman** [comment]
You're missing a few components I consider essential to reliably testing ansible code.

* Molecule
* Virtualenv + `pip freeze` to pin Python dependencies.

I've thoroughly documented how to go about doing this in the GIMP CI docs.  https://github.com/gimp-ci/documentation/blob/master/creating-ansible-roles.md

Here's an example role - https://github.com/gimp-ci/ansible-gimp-dependencies

**Comment by mattyjones** [comment]
@samrocketman Thanks for the comments, this is a work in progress as I find time during the week. I will certainly take your thoughts to heart.

I will say that I am torn about pinning deps. Normally in a production environment I would say pin everything but in the case of this if I am building against latest that won't change much in Debian land. I am not against pinning by no means, just unsure of the value here.

Molecule is a tool I am unfamiliar with so I will read up on that when I get a few minutes. Thanks for the links as well!

**Comment by phoerious** [comment]
I have no special opinion on Ansible. As long as you think it's helpful, go for it.
We build our AppImage using our release-tool and a simple Docker container, so we're basically doing something similar, but with a little less provisioning overhead.

**Comment by mattyjones** [comment]
@samrocketman Just spun through using molecule and testinfra (instead of serverspec). From the outset I like them, I will keep them in mind as I go forward. I have a lot vested in serverspec coming from using Chef professionally but I personally have no love for Chef or Ruby. Python was my first goto language and I much prefer yaml/toml to most anything else.

Thanks for the tips, well worth the read and time taken to get to know the tools.

**Comment by samrocketman** [comment]
@mattyjones glad you found them useful.  For me, the biggest thing is the `virtualenv` tool with `pip freeze` which can be used to version the exact version of all of the python tools regardless of the OS being used (akin to `Gemfile.lock` in Ruby).

**Comment by samrocketman** [comment]
@phoerious Ansible + molecule can be used to provision and test docker containers as well.  It's more about testing infra and repeatability.

**Comment by phoerious** [comment]
Tell us when you're at a point where you think we could do some test builds. If you want, you can also join us on IRC, so we can talk everything through.
Do you have any specific plans how and when you want to submit the package to Debian?

**Comment by mattyjones** [comment]
@phoerious 

I will join on IRC when I have some time maybe the beginning of the week. Monday I will be afk much of the day. 

## Testing Builds

I have already built the software itself using the Debian role without issue. It is worth noting though that I do not include any X packages in the role nor do I list X as a dependency. This is standard Debian policy that I clarified earlier this week. I agree with it as well due to the simple fact that you can, and indeed many people do, forward their X server. FWIW when you build the software without X and then try to run it, it will simply say `No X server found` which IMHO is perfectly legit. Requiring QT brings with it the implicit expectation that X is also required.

## Package Planning

I have no timetable on when I will have official packages ready to submit. The best guess is, when they are ready. ;) I am hoping within 30 days, honestly though it could be 2 weeks till the first rough package is ready and I shop around for a sponsor. When it comes to how I will do it, the path is much more clear. I have already taken ownership of the [bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=855173) attached to this project and changed the title to **ITP** (Intent To Publish). When I get a package installed that meets the quality and additional guidelines needed I will then request a Sponsor for the package who will then verify it, sign it, and upload it themselves.

I will then continue to be the active community maintainer of this package, taking responsibility for all packaging related bugs/issues as well as other bugs (as time allows). At some point, when I know I can commit the time and resources needed, I will apply to be an official Debian maintainer and then fully adopt this package as mine w/o the need for a sponsor to upload new versions.

**Comment by phoerious** [comment]
That sounds amazing. 

One thing about X, though: KeePassXC depends directly on libxi and libxtst for Auto-Type. Please verify that this is working.

**Comment by droidmonkey** [comment]
Recommend rebasing this ontop of new master. Also, any progress?

File: .gitignore
Comment (samrocketman) [inline .gitignore:7]: Consider adding `*.pyc`.

**Review by samrocketman — CHANGES_REQUESTED** [review]
Two binaries are included with this change and should be removed:
* `roles/common/tests/__pycache__/test_build.cpython-27-PYTEST.pyc`
* `roles/common/tests/__pycache__/test_keypassxc.cpython-27-PYTEST.pyc`

**Comment by TheZ3ro** [comment]
Progress here?

**Comment by hifi** [comment]
Spoke with @LocutusOfBorg and he would be willing to be a sponsor to upload the Debian package. However, he would rather prefer if someone with more personal interest would pick this up. His instructions were that someone should start a thread on debian-devel to potentially get a more involved sponsor.

Hope this helps to get the packaging effort forward!

**Comment by mattyjones** [comment]
Sorry for the drop off on this a job change and personal issues negated all open source commitment. I will be pushing this through by the end of the week and will hopefully have an initial package ready as well.

I will start the thread and also I am in the process of becoming a maintainer and will assume responsibility for this project.

**Comment by mattyjones** [comment]
@phoerious This is ready for merging as the provisioner is all set. The next thing will be a new pr for the actual building of the package. After that it will just be cleanups and then figure out if we want to automate the process using a secure vagrant environment and such for any other bits. I am specifically thinking signed rpms and tarballs.

**Comment by samrocketman** [comment]
@mattyjones mind squashing this since it's ready to merge?  (note I'm not a maintainer and don't know if squashing is what is desired from the maintainers).

**Comment by droidmonkey** [comment]
Does this come with an instruction manual?

**Comment by mattyjones** [comment]
@droidmonkey nice point! I will add something in the readme for now, but was looking at something in the main docs. The goal would be for anyone to independently verify the build process. Basically do everything except sign it with the projects pvt key.

**Pull request closed without merging**
